[GStreamer] Allow RegistryScanner call sites to use the selected element factory

philn opened the pull request:
#### e3ba18ed2f68f5085f83e1cfedfbbcec2f9ab805
<pre>
[GStreamer] Allow RegistryScanner call sites to use the selected element factory
<a href="https://bugs.webkit.org/show_bug.cgi?id=248147">https://bugs.webkit.org/show_bug.cgi?id=248147</a>

Reviewed by Xabier Rodriguez-Calvar.

When a candidate GstElementFactory has been selected by the scanner, it can now be used to create a
GstElement matching the capabilies that were probed. This can be useful in cases where a higher
level bin such as decodebin or encodebin is not desired and we only need direct access to the
element.

* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::ElementFactories::hasElementForMediaType const):
(WebCore::GStreamerRegistryScanner::fillMimeTypeSetFromCapsMapping):
(WebCore::GStreamerRegistryScanner::initializeDecoders):
(WebCore::GStreamerRegistryScanner::initializeEncoders):
(WebCore::GStreamerRegistryScanner::isCodecSupported const):
(WebCore::GStreamerRegistryScanner::isAVC1CodecSupported const):
(WebCore::GStreamerRegistryScanner::isConfigurationSupported const):
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.h:
(WebCore::GStreamerRegistryScanner::RegistryLookupResult::merge):
(WebCore::GStreamerRegistryScanner::CodecLookupResult::CodecLookupResult):
(WebCore::GStreamerRegistryScanner::CodecLookupResult::operator bool const):

Canonical link: <a href="https://commits.webkit.org/256944@main">https://commits.webkit.org/256944@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/31a59b5935c510eef03829305abc94a3b871d52e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97315 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6579 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30471 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106834 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167096 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101282 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6880 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35316 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89725 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103513 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102978 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83944 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32164 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/87019 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/595 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/578 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4784 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5378 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1822 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41104 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
```diff
@@ -279,27 +286,27 @@ void GStreamerRegistryScanner::fillMimeTypeSetFromCapsMapping(const GStreamerReg
-        m_decoderCodecMap.add(AtomString("mpeg"_s), false);
-        m_decoderCodecMap.add(AtomString("mp4a*"_s), false);
+        m_decoderCodecMap.add(AtomString("mpeg"_s), result);
+        m_decoderCodecMap.add(AtomString("mp4a*"_s), result);
```
this is always going to be true because if result is false, this is never entering the if branch. Might we be missing anything?

Ditto for all if branches below.

philn replied:
That `false` meant "not using hardware acceleration". `result` having its bool operator return true means we have an element matching the media type. We now additionally store the result in the codec map, that's all, I don't think this is a change of behavior.

calvaris replied:
What I mean is that you're passing a result that is true, otherwise you would not have entered that if branch, as result is checked and has to be true to enter that if branch, right?

The case of other branches below is a bit different because you passed before the hardware support, but now you also pass a variable that you check in the if branch, being always true.

philn replied:
What is the problem with that?

philn replied:
The result evaluates to `true` but is a struct...

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.h:
```diff
@@ -79,7 +81,21 @@ class GStreamerRegistryScanner {
-    bool isCodecSupported(Configuration, const String& codec, bool usingHardware = false) const;
+    struct CodecLookupResult {
+        CodecLookupResult() = default;
+        CodecLookupResult(bool isSupported, GRefPtr<GstElementFactory> factory)
```
I think we should either use `const GRefPtr&` or `GRefPtr&&`, or and overload with both depending on the use case.

philn replied:
I'll try the former. Can't be the latter because it's common to add the same factory to multiple `CodecLookupResult's.

calvaris reviewed (approved):
With the constructor change, you're good to go

philn committed:
[GStreamer] Allow RegistryScanner call sites to use the selected element factory
https://bugs.webkit.org/show_bug.cgi?id=248147

Reviewed by Xabier Rodriguez-Calvar.

When a candidate GstElementFactory has been selected by the scanner, it can now be used to create a
GstElement matching the capabilies that were probed. This can be useful in cases where a higher
level bin such as decodebin or encodebin is not desired and we only need direct access to the
element.

* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::ElementFactories::hasElementForMediaType const):
(WebCore::GStreamerRegistryScanner::fillMimeTypeSetFromCapsMapping):
(WebCore::GStreamerRegistryScanner::initializeDecoders):
(WebCore::GStreamerRegistryScanner::initializeEncoders):
(WebCore::GStreamerRegistryScanner::isCodecSupported const):
(WebCore::GStreamerRegistryScanner::isAVC1CodecSupported const):
(WebCore::GStreamerRegistryScanner::isConfigurationSupported const):
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.h:
(WebCore::GStreamerRegistryScanner::RegistryLookupResult::merge):
(WebCore::GStreamerRegistryScanner::CodecLookupResult::CodecLookupResult):
(WebCore::GStreamerRegistryScanner::CodecLookupResult::operator bool const):

Canonical link: https://commits.webkit.org/256944@main
